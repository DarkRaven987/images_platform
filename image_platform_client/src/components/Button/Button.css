.button-item {
  background-color: rgb(66, 170, 255);
  color: white;
  padding: 10px 15px;
  font-size: 14px;
  font-weight: bold;
  border-radius: 8px;
  border: none;
  cursor: pointer;
  box-shadow: 0 0 0 white;
  transition: ease-in-out 0.3s;
}

.button-item:hover {
  background-color: rgb(39, 54, 255);
  box-shadow: 0px 2px 4px rgb(66, 170, 255);
}
