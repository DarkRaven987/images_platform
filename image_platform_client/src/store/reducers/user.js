import { createSlice } from '@reduxjs/toolkit';
import { agent } from '../../utils/agent';
import { API_SIGN_IN_URL, API_SIGN_UP_URL } from '../../utils/agentConsts';

const initialState = null;

const { actions, reducer } = createSlice({
  name: 'user',
  initialState,
  reducers: {
    saveUserData(state, action) {
      return (state = action.payload);
    },
    clearUserData(state, action) {
      return (state = action.payload);
    },
  },
});

export const { saveUserData, clearUserData } = actions;

export const signInAction = async ({ dispatch, username, password }) => {
  if (!username || !password) return false;

  try {
    const userData = await agent.post(API_SIGN_IN_URL, { username, password });
    const { user, accessToken, refreshToken } = userData?.data;

    if (!user?.PK) return false;

    dispatch(saveUserData(user));
    localStorage.setItem('user', JSON.stringify(user));
    localStorage.setItem('accessToken', `Bearer ${accessToken}`);
    localStorage.setItem('refreshToken', `Bearer ${refreshToken}`);

    return true;
  } catch (err) {
    console.error(err);
    return false;
  }
};

export const signUpAction = async ({ username, password }) => {
  if (!username || !password) return false;

  try {
    await agent.post(API_SIGN_UP_URL, { username, password });

    return true;
  } catch (err) {
    console.error(err);
    return false;
  }
};

export const signOutAction = () => {
  localStorage.removeItem('user');
  localStorage.removeItem('accessToken');
  localStorage.removeItem('refreshToken');
  window.location.replace(`${window.location.origin}/login`);
};

export default reducer;
