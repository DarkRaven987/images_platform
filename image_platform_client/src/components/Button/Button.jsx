import './Button.css';

const Button = ({ children, className = '', type = 'button', ...props }) => {
  return (
    <button
      data-testid="button-test"
      className={`button-item ${className}`}
      type={type}
      {...props}
    >
      {children}
    </button>
  );
};

export default Button;
