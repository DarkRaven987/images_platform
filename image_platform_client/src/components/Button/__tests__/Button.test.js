import { render, screen, cleanup } from '@testing-library/react';
import user from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Button from '../Button';

const innerLabel = 'Testing button';

describe('Button component:', () => {
  afterEach(() => {
    cleanup();
  });

  it('* renders successfully', () => {
    render(<Button>{innerLabel}</Button>);
    const ButtonElement = screen.getByTestId('button-test');
    expect(ButtonElement).toBeTruthy();
    expect(ButtonElement.textContent).toBe(innerLabel);
  });

  it('* accepts custom classes', () => {
    const customClassName = 'custom-button';
    render(<Button className={customClassName}>{innerLabel}</Button>);
    const ButtonElement = screen.getByTestId('button-test');
    expect(ButtonElement.classList).toContain(customClassName);
  });

  it('* accepts button type attribute', () => {
    render(<Button type="submit">{innerLabel}</Button>);
    const ButtonElement = screen.getByTestId('button-test');
    expect(ButtonElement.type).toBe('submit');
  });

  it('* accepts custom properties', () => {
    const test = renderer
      .create(
        <Button type="button" data-attr1="attr-1" data-attr2="attr-2">
          {innerLabel}
        </Button>,
      )
      .toJSON();
    expect(test.props['data-attr1']).toBe('attr-1');
    expect(test.props['data-attr2']).toBe('attr-2');
    expect(test.props.type).toBe('button');
  });

  it('* matches the snapshot', () => {
    render(
      <Button type="submit" className="snapshot-btn-class">
        {innerLabel}
      </Button>,
    );
    const ButtonElement = screen.getByTestId('button-test');
    expect(ButtonElement).toMatchSnapshot('test-button-snapshot');
  });

  it('* button is clickable', async () => {
    user.setup();

    let test = 0;

    render(
      <Button
        type="submit"
        className="snapshot-btn-class"
        onClick={() => (test += 1)}
      >
        {innerLabel}
      </Button>,
    );
    const ButtonElement = screen.getByTestId('button-test');

    expect(test).toBe(0);
    await user.click(ButtonElement);
    expect(test).toBe(1);
    await user.click(ButtonElement);
    await user.click(ButtonElement);
    expect(test).toBe(3);
  });

  it('* button is disabled and not clickable', async () => {
    user.setup();

    let test = 0;

    render(
      <Button
        type="submit"
        className="snapshot-btn-class"
        onClick={() => (test += 1)}
        disabled={true}
      >
        {innerLabel}
      </Button>,
    );
    const ButtonElement = screen.getByTestId('button-test');

    expect(ButtonElement.disabled).toBe(true);

    expect(test).toBe(0);
    await user.click(ButtonElement);
    expect(test).not.toBe(1);
    await user.click(ButtonElement);
    await user.click(ButtonElement);
    expect(test).not.toBe(3);
  });
});
