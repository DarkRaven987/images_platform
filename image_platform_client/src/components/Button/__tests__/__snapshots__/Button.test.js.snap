// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component: * matches the snapshot: test-button-snapshot 1`] = `
<button
  class="button-item snapshot-btn-class"
  data-testid="button-test"
  type="submit"
>
  Testing button
</button>
`;
